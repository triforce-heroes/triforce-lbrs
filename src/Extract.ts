import { BufferConsumer } from "@triforce-heroes/triforce-core/BufferConsumer";

import type { Entry } from "@/types/Entry";

export function extract(data: Buffer) {
  const consumer = new BufferConsumer(data, 16);

  const fileId = consumer.readUnsignedInt32();
  const pathId = consumer.readUnsignedInt32();

  const entries: Entry[] = [];
  const entriesCount = consumer.skip(16).readUnsignedInt32();

  for (let i = 0; i < entriesCount; i++) {
    const name = consumer.readLengthPrefixedString();

    const translationsCount = consumer.skipPadding(4).readUnsignedInt32();
    const translations: string[] = [];

    for (let j = 0; j < translationsCount; j++) {
      translations.push(consumer.readLengthPrefixedString());
      consumer.skipPadding(4);
    }

    entries.push([name, translations]);
  }

  return { fileId, pathId, entries };
}
