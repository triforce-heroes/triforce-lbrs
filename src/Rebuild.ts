import { BufferBuilder } from "@triforce-heroes/triforce-core/BufferBuilder";

import type { Entry } from "@/types/Entry";

export function rebuild(fileId: number, pathId: number, entries: Entry[]) {
  const builder = new BufferBuilder();

  // Header #1.
  builder.writeUnsignedInt32(0);
  builder.writeUnsignedInt32(0);
  builder.writeUnsignedInt32(0);
  builder.writeUnsignedInt32(1);

  // Header #2.
  builder.writeUnsignedInt32(fileId);
  builder.writeUnsignedInt32(pathId);
  builder.writeUnsignedInt32(0);
  builder.writeLengthPrefixedString("texts_db");

  // Entries.
  builder.writeUnsignedInt32(entries.length);

  for (const [name, translations] of entries) {
    builder.writeLengthPrefixedString(name);
    builder.pad(4);

    builder.writeUnsignedInt32(translations.length);

    for (const translation of translations) {
      builder.writeLengthPrefixedString(translation);
      builder.pad(4);
    }
  }

  return builder.build();
}
