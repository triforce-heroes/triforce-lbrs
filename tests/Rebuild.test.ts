import { describe, expect, it } from "vitest";

import { rebuild } from "@/Rebuild.js";

import type { Entry } from "@/types/Entry";

import { extract } from "@/Extract";

describe("rebuild() function", () => {
  type Test = [fileId: number, pathId: number, entries: Entry[]];

  const tests: Test[] = [
    [0, 1, []],
    [0, 2, [["HELLO", ["W", "O", "R", "L", "D", "", "!"]]]],
    [
      0,
      3,
      [
        ["A", ["B", "C"]],
        ["D", []],
      ],
    ],
    [0, 4, [["Hello", ["World!"]]]],
    [
      0,
      5,
      [
        ["Hello", ["World!"]],
        ["Hello", ["World!"]],
      ],
    ],
    [
      0,
      6,
      [
        ["ABCD", ["ABCD", "", "1234"]],
        ["123456", ["ABCD", "", "1234"]],
      ],
    ],
  ];

  it.each(tests)("[#%#] rebuild(%j, %j, ...)", (fileId, pathId, entries) => {
    const rebuilded = rebuild(fileId, pathId, entries);

    expect(rebuilded).toMatchSnapshot();
    expect(extract(rebuilded)).toStrictEqual({ fileId, pathId, entries });
  });
});
