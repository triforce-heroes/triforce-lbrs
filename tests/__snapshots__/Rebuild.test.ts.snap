// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rebuild() function > [#0] rebuild(0, 1, ...) 1`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    8,
    0,
    0,
    0,
    116,
    101,
    120,
    116,
    115,
    95,
    100,
    98,
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`rebuild() function > [#1] rebuild(0, 2, ...) 1`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    2,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    8,
    0,
    0,
    0,
    116,
    101,
    120,
    116,
    115,
    95,
    100,
    98,
    1,
    0,
    0,
    0,
    5,
    0,
    0,
    0,
    72,
    69,
    76,
    76,
    79,
    0,
    0,
    0,
    7,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    87,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    79,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    82,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    76,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    68,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    33,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`rebuild() function > [#2] rebuild(0, 3, ...) 1`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    3,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    8,
    0,
    0,
    0,
    116,
    101,
    120,
    116,
    115,
    95,
    100,
    98,
    2,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    65,
    0,
    0,
    0,
    2,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    66,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    67,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    68,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`rebuild() function > [#3] rebuild(0, 4, ...) 1`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    4,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    8,
    0,
    0,
    0,
    116,
    101,
    120,
    116,
    115,
    95,
    100,
    98,
    1,
    0,
    0,
    0,
    5,
    0,
    0,
    0,
    72,
    101,
    108,
    108,
    111,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    6,
    0,
    0,
    0,
    87,
    111,
    114,
    108,
    100,
    33,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`rebuild() function > [#4] rebuild(0, 5, ...) 1`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    5,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    8,
    0,
    0,
    0,
    116,
    101,
    120,
    116,
    115,
    95,
    100,
    98,
    2,
    0,
    0,
    0,
    5,
    0,
    0,
    0,
    72,
    101,
    108,
    108,
    111,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    6,
    0,
    0,
    0,
    87,
    111,
    114,
    108,
    100,
    33,
    0,
    0,
    5,
    0,
    0,
    0,
    72,
    101,
    108,
    108,
    111,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    6,
    0,
    0,
    0,
    87,
    111,
    114,
    108,
    100,
    33,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`rebuild() function > [#5] rebuild(0, 6, ...) 1`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    6,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    8,
    0,
    0,
    0,
    116,
    101,
    120,
    116,
    115,
    95,
    100,
    98,
    2,
    0,
    0,
    0,
    4,
    0,
    0,
    0,
    65,
    66,
    67,
    68,
    3,
    0,
    0,
    0,
    4,
    0,
    0,
    0,
    65,
    66,
    67,
    68,
    0,
    0,
    0,
    0,
    4,
    0,
    0,
    0,
    49,
    50,
    51,
    52,
    6,
    0,
    0,
    0,
    49,
    50,
    51,
    52,
    53,
    54,
    0,
    0,
    3,
    0,
    0,
    0,
    4,
    0,
    0,
    0,
    65,
    66,
    67,
    68,
    0,
    0,
    0,
    0,
    4,
    0,
    0,
    0,
    49,
    50,
    51,
    52,
  ],
  "type": "Buffer",
}
`;
